### PR TITLE
New WebGL Exporter

### PR DESCRIPTION
**Pull request by travisapple**
This is a complete rewrite of this file.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [ ] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Comment by luzpaz** [comment]
Uhmm... why is Yorik deleted from the acknowledgments? That's not how it's done. Even if the code has been rewritten one still acknowledges the previous authors.

**Comment by travisapple** [comment]
It's a complete rewrite. All new. This is my first time doing this so I'm not sure about the etiquette required for acknowledgments.

File: src/Mod/Arch/importWebGL.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #***************************************************************************
-#*   Copyright (c) 2013 Yorik van Havre <yorik@uncreated.net>              *
+#*   Copyright (c) 2020 Travis Apple <travisapple@gmail.com>               *
```
Comment (luzpaz) [inline src/Mod/Arch/importWebGL.py:2]: Please keep the reference to yorik

**Review by luzpaz — CHANGES_REQUESTED** [review]

File: src/Mod/Arch/importWebGL.py [hunk 2]
```diff
@@ -18,233 +19,782 @@
...
+# https://forum.freecadweb.org/viewtopic.php?t=29487
+# https://threejs.org/examples/#webgl_raycast_sprite
+#
+# Params for export()
+#    'colors' is of the form: {'Body': [1,0,0], 'Body001': [1,1,0], 'Body002': [1,0,1] }
+#    'camera' is of the form: "PerspectiveCamera {\n  viewportMapping ADJUST_CAMERA\n  position 30.242626 -51.772324 85.63475\n  orientation -0.4146691 0.088459305 -0.90566254  4.7065201\nnearDistance 53.126431\n  farDistance 123.09125\n  aspectRatio 1\n  focalDistance 104.53851\n  heightAngle 0.78539819\n\n}"
+#    The 'camera' string for the active document may be generated from: import OfflineRenderingUtils; OfflineRenderingUtils.getCamera(FreeCAD.ActiveDocument.FileName);
+#
+# Development reload oneliner:
+# def re(): from importlib import reload;import importWebGL;reload(importWebGL);o=FreeCAD.getDocument("curve");importWebGL.export([o.getObject("Body")],u"C:/Users/Travis/Desktop/test.htm");
```
Comment (luzpaz) [inline src/Mod/Arch/importWebGL.py:36]: `C:/Users/Travis/Desktop/test.htm"` is hardcoded

File: src/Mod/Arch/importWebGL.py [hunk 3]
```diff
@@ -18,233 +19,782 @@
...
-    def translate(ctxt,txt,utf8_decode=True):
-        return txt
-    # \endcond
-
-## @package importWebGL
-#  \ingroup ARCH
-#  \brief WebGL file format exporter
-#
-#  This module provides tools to export HTML files containing the 
-#  exported objects in WebGL format and a simple three.js-based viewer.
```
Comment (luzpaz) [inline src/Mod/Arch/importWebGL.py:46]: Doxygen comments were removed but never replaced

File: src/Mod/Arch/importWebGL.py (hunk 2 above)
Comment (travisapple) [reply src/Mod/Arch/importWebGL.py:36]: That's just an example for the next dev. It took me a while to find a quick way to test this code and so I put that in there to save a step for them.

File: src/Mod/Arch/importWebGL.py (hunk 3 above)
Comment (travisapple) [reply src/Mod/Arch/importWebGL.py]: This has been added back in

File: src/Mod/Arch/importWebGL.py (hunk 2 above)
Comment (travisapple) [reply src/Mod/Arch/importWebGL.py]: But you're right. I've replaced the names with something more descriptive

**Comment by luzpaz** [comment]
FYI, this is why it's better to make a PR that shows several commits explaining what you're adding in the code sequentially. Instead of submitting one giant re-written file chunked all in one commit. It makes it harder to review.

**Comment by travisapple** [comment]
Three.js has changed so much in the years since this was written that there was no way this could have been rewritten bit by bit.

**Comment by yorikvanhavre** [comment]
Looks great! Indeed this was in dire need of update... Will give it a test spin ASAP before merging.

**Comment by travisapple** [comment]
Thank you Yorik. I'm grateful for any feedback you can give me, especially in the area around where it says "Sub-Links are Unsupported". I was unable to figure out how to handle the placement of sub-links because there seems to be two concurrent placements that need to be added together and then applied.

**Comment by X-Ryl669** [comment]
First congratulation for your work, the previous version was so... old.
I was wondering why you did not move to export to GLTF directly ? 4 years ago it was too new, but now it's the current state of art.
If FreeCAD is upgraded to use the latest OpenCascade, then GLTF export is provided for almost *free*. It keeps the scene, the materials and the camera in the output, thus the final user doesn't have to reposition itself hardly.
I made a small demo last year, and I think it's much more usable that the previous Webgl export.

Did you see this code: https://github.com/X-Ryl669/freecadWebExport ?

**Comment by travisapple** [comment]
I had not seen this and it looks great! I will certainly use it on my own projects in the near future and I appreciate you bring this up here. I agree that GLTF has come a looooong way.

However, the reason I'm rewriting this exporter is that I'm trying to move from F360 and I need a **quick** way to email an updated design to a client that is entirely self contained and that I've got a 99% chance of it working on their (probably trash) computer/phone. F360 has this and it works great and FC does not and is causing me pain.

This code relies on Three.js and that has proven to be almost bulletproof on all kinds of hardware. It's entirely self-contained, meaning that I can just email a file to the client. I do not need to upload anything to a web server if I don't want to. 15 seconds and I'm done.

**Comment by yorikvanhavre** [comment]
I actually cannot make it work? I created a simple test file with a default part cube and cylinder, and exported the two objects, but nothing appears in the viewer...

**Comment by travisapple** [comment]
After looking at this for way too long, it was my testing method that was flawed. You loaded a couple objects in, clicked on them and then exported, as any user would. However, this makes Arch.pruneIncluded() reject that object.

I've remove that line. Please try again.

**Comment by yorikvanhavre** [comment]
Arch.pruneIncluded() should not remove these two objects normally... I'll check on that and fix where needed (and we can then revert your last commit if needed)

**Comment by yorikvanhavre** [comment]
Great, thanks! This seems to work okay. But I see another problem:

1. Create/open a document
2. Export it
3. Create/open another document
4. Export it

Result: the second exported file contains the objects from the first document, not the second

**Comment by travisapple** [comment]
I see it. I'm looking at importDAE.py and import.IFC.py (you wrote these) for reference and I can't see any difference in how exportList is handled, yet these export correctly. Is there some internal caching here that I can turn off?

Running the following lines in the console between exports fixes this issue:
```
from importlib import reload
import importWebGL
reload(importWebGL)
```

Is my code somehow not releasing exportList on the first export and stopping it from being changed? I'm sorry Yorik, I just don't understand what's happening here.

**Comment by yorikvanhavre** [comment]
Ok no worries, I'll have a look at it too... Normally you don't need to do anything else than handling exportList, but maybe some global variable is holding stuff somewhere, or a new bug arrived in FreeCAD :)

**Comment by X-Ryl669** [comment]
BTW, I've reworked my demo so it runs with a [single HTML file](https://github.com/X-Ryl669/freecadWebExport/tree/singleFile) as you mentioned. It's still not as good as F360 web viewer, but at least it's simple and open source.

**Comment by travisapple** [comment]
> 
> 
> Ok no worries, I'll have a look at it too... Normally you don't need to do anything else than handling exportList, but maybe some global variable is holding stuff somewhere, or a new bug arrived in FreeCAD :)

Thank you Yorik. I felt like I hit a dead end there.

**Comment by travisapple** [comment]
> 
> 
> BTW, I've reworked my demo so it runs with a [single HTML file](https://github.com/X-Ryl669/freecadWebExport/tree/singleFile) as you mentioned. It's still not as good as F360 web viewer, but at least it's simple and open source.

Very cool! I was unable to export to glTF using the new exporter because OCCT 7.5 is required but I'll take a look tonight. Do you know if there are plans to not require this in the future?

**Comment by X-Ryl669** [comment]
None that I'm aware of. GLTF exporter from OCCT is used and you need OCCT version 7.5 to have it (and it must be built with RapidJSON dependency).

**Comment by yorikvanhavre** [comment]
Ok I think the problem is just that html is declared as global at L567. So, later on, when you modify it, it modifies the template itself and not a copy of that template. So in subsequent runs, there is no more $data to replace...
The solution is probably simply to remove html from L567...

**Comment by travisapple** [comment]
Yep, that did the trick. I'm sorry Yorik, I didn't realize that we're not running the script as it's written every time. We're loading the script and running export() every time. Thanks for your help on this.

**Comment by yorikvanhavre** [comment]
Thanks! Now there is one problem left, the file has changed so much that github is unable to reconciliate it with the version currently in master branch. Two solutions: 1) You rebase it yourself (there will be many >>>> and <<<< to solve), or 2) I can simply overwrite the file with your version here, but then the commit will be made by me and not by you, and you deserve the credit ;) So I'd go for 1) if you have the patience to do it, otherwise tell me and I'll do 2)

**Comment by travisapple** [comment]
This is my first time through with this. I **think** that did it.

**Comment by yorikvanhavre** [comment]
Ok all done, thanks! Merging now

**Comment by stillpointx** [comment]
Getting error **"gITF support requires OCCT 7.5.0 or later"**
```
05:10:04  <Import> ImportOCAF2.cpp(1134): Unnamed#untitled_Hex_Holder_I3Mk3 has null shape
05:10:04  Traceback (most recent call last):
  File "<string>", line 4, in <module>
<class 'RuntimeError'>: gITF support requires OCCT 7.5.0 or later

```
System Used:
OS: Ubuntu 20.04.2 LTS (ubuntu:GNOME/ubuntu)
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.24276 (Git) AppImage
Build type: Release
Branch: (HEAD detached at 0.19.1)
Hash: a88db11e0a908f6e38f92bfc5187b13ebe470438
Python version: 3.8.8
Qt version: 5.12.9
Coin version: 4.0.0
OCC version: 7.4.0
Locale: English/United States (en_US)

**Comment by travisapple** [comment]
> 
> 
> Getting error **"gITF support requires OCCT 7.5.0 or later"**
> 
> ```
> 05:10:04  <Import> ImportOCAF2.cpp(1134): Unnamed#untitled_Hex_Holder_I3Mk3 has null shape
> 05:10:04  Traceback (most recent call last):
>   File "<string>", line 4, in <module>
> <class 'RuntimeError'>: gITF support requires OCCT 7.5.0 or later
> ```
> 
> System Used:
> OS: Ubuntu 20.04.2 LTS (ubuntu:GNOME/ubuntu)
> Word size of OS: 64-bit
> Word size of FreeCAD: 64-bit
> Version: 0.19.24276 (Git) AppImage
> Build type: Release
> Branch: (HEAD detached at 0.19.1)
> Hash: [a88db11](https://github.com/FreeCAD/FreeCAD/commit/a88db11e0a908f6e38f92bfc5187b13ebe470438)
> Python version: 3.8.8
> Qt version: 5.12.9
> Coin version: 4.0.0
> OCC version: 7.4.0
> Locale: English/United States (en_US)

This thread was for is the WebGL exporter and not the glTF exporter

**Pull request merged**